File: src/sliding-windows/substring-anagram.ts
export function findAnagrams(s: string, target: string): number[] {
    const result: number[] = []

    if (s.length < target.length) {
        return result
    }

    const targetCharFreq = new Array(26).fill(0)
    for (let i = 0; i < target.length; i++) {
        targetCharFreq[target.charCodeAt(i) - 'a'.charCodeAt(0)]++
    }
    const windowCharFreq: number[] = new Array(26).fill(0)
    const isFreqEqual = () => {
        for (let i = 0; i < targetCharFreq.length; i++) {
            if (targetCharFreq[i] !== windowCharFreq[i]) {
                return false
            }
        }
        return true
    }

    let start = 0,
        end = 0
    while (end < s.length) {
        windowCharFreq[s.charCodeAt(end) - 'a'.charCodeAt(0)]++

        if (end - start + 1 === target.length) {
            if (isFreqEqual()) {
                result.push(start)
            }
            windowCharFreq[s.charCodeAt(start) - 'a'.charCodeAt(0)]--
            start++
        }

        end++
    }

    return result
}

File: src/sliding-windows/substring_anagram.go
package slidingwindows

func FindAnagram(s string, target string) []int {
	result := make([]int, 0)

	if len(s) < len(target) {
		return result
	}

	targetCharFreq := make([]int, 26)
	for i := 0; i < len(target); i++ {
		targetCharFreq[target[i]-'a']++
	}
	windowCharFreq := make([]int, 26)
	isFreqEqual := func() bool {
		for i := 0; i < len(targetCharFreq); i++ {
			if targetCharFreq[i] != windowCharFreq[i] {
				return false
			}
		}
		return true
	}

	start := 0
	for end := 0; end < len(s); end++ {
		windowCharFreq[s[end]-'a']++

		if end-start+1 == len(target) {
			if isFreqEqual() {
				result = append(result, start)
			}
			windowCharFreq[s[start]-'a']--
			start++
		}
	}

	return result
}
